File: cpp/src/stream.cpp
#ifdef GBK
#include <iconv.h>
#endif

#include <spdlog/spdlog.h>
#include <cpprest/http_client.h>
#include <cpprest/streams.h>
#include <zlib.h>
#include <filesystem>
#include <optional>
#include <atomic>


struct ZipHeader {
    std::uint16_t version;
    std::uint16_t flags;
    std::uint16_t compression;
    std::uint16_t mod_time;
    std::uint16_t mod_date;
    std::uint32_t crc32_expected;
    std::uint32_t compressed_size;
    std::uint32_t uncompressed_size;
    std::uint16_t file_name_length;
    std::uint16_t extra_field_length;
};

template<typename T>
T readFromByteArray(const std::byte* &ptr) {
    T value;
    std::memcpy(&value, ptr, sizeof(T));
    ptr += sizeof(T);
    return value;
}

[[nodiscard]] ZipHeader parseHeader(const std::byte* headerData) {
    ZipHeader header;
    header.version              = readFromByteArray<std::uint16_t>(headerData);
    header.flags                = readFromByteArray<std::uint16_t>(headerData);
    header.compression          = readFromByteArray<std::uint16_t>(headerData);
    header.mod_time             = readFromByteArray<std::uint16_t>(headerData);
    header.mod_date             = readFromByteArray<std::uint16_t>(headerData);
    header.crc32_expected       = readFromByteArray<std::uint32_t>(headerData);
    header.compressed_size      = readFromByteArray<std::uint32_t>(headerData);
    header.uncompressed_size    = readFromByteArray<std::uint32_t>(headerData);
    header.file_name_length     = readFromByteArray<std::uint16_t>(headerData);
    header.extra_field_length   = readFromByteArray<std::uint16_t>(headerData);
    return header;
}


class ZipContentFetcher {
public:
    web::http::http_response currentResponse;
    std::atomic<size_t> totalSize;
    std::atomic<size_t> progress;
    std::chrono::steady_clock::time_point startTime;

    ZipContentFetcher(const std::vector<std::string>& urls) : urls(urls), currentUrlIndex(0), totalSize(0), progress(0) {
        startTime = std::chrono::steady_clock::now();
        switchStream();
    }

    void switchStream() {
        if (currentUrlIndex >= urls.size()) {
            spdlog::info("Processing complete");
            exit(0);
        }

        const auto url = urls[currentUrlIndex++];
        web::http::client::http_client client(utility::conversions::to_string_t(url));

        client.request(web::http::methods::GET, U("/")).then([this, url](const web::http::http_response response) {
            if (response.status_code() == web::http::status_codes::OK) {
                this->currentResponse = response;
                totalSize = response.headers().content_length();
                spdlog::info("Processing volume: {}", url);
                progress = 0;
                startTime = std::chrono::steady_clock::now();
            } else {
                spdlog::info("Request failed with status code: {}", response.status_code());
                switchStream();
            }
        }).wait();
    }
    
    [[nodiscard]] const std::optional<std::vector<std::byte>> read(const size_t length) {
        std::vector<std::byte> data(length);
        auto buf = currentResponse.body().streambuf();
        const auto bytesRead = buf.getn(reinterpret_cast<uint8_t*>(data.data()), length).get();
        updateProgress(bytesRead);

        if (bytesRead < length) {
            switchStream();
            auto remainingData = read(length - bytesRead);
            if (!remainingData) {
                return std::nullopt;
            }
            data.insert(data.end(), remainingData->begin(), remainingData->end());
        }

        data.resize(bytesRead);
        return data;
    }
private:
    std::vector<std::string> urls;
    size_t currentUrlIndex;

    void updateProgress(const size_t bytesRead) noexcept {
        progress += bytesRead;
        displayProgress();
    }

    void displayProgress() noexcept {
        auto now = std::chrono::steady_clock::now();
        auto timeElapsed = std::chrono::duration_cast<std::chrono::seconds>(now - startTime).count();
        auto speed = timeElapsed > 0 ? (progress / 1024.0 / 1024.0) / timeElapsed : 0; // MB/s

        int barWidth = 70;
        float progressPercentage = totalSize > 0 ? (float)progress / totalSize : 0.0;
        std::cout << "[";
        int pos = barWidth * progressPercentage;
        for (int i = 0; i < barWidth; ++i) {
            if (i < pos) std::cout << "=";
            else if (i == pos) std::cout << ">";
            else std::cout << " ";
        }
        std::cout << "] " << std::setprecision(2) << int(progressPercentage * 100.0) << " % " << speed << " MB/s\r";
        std::cout.flush();
    }
};

#ifdef GBK
[[nodiscard]] std::string convertBytesToString(const std::vector<std::byte>& gbkData) {
    size_t inBytes = gbkData.size();
    char* inputPtr = const_cast<char*>(reinterpret_cast<const char*>(gbkData.data()));
    
    size_t outBytes = 4 * gbkData.size();
    std::vector<char> utf8Data(outBytes);
    char* outputPtr = utf8Data.data();
    
    const iconv_t cd = iconv_open("UTF-8", "GBK");
    if (cd == (iconv_t)-1) {
        throw std::runtime_error("Failed to open iconv");
    }
    
    const size_t result = iconv(cd, &inputPtr, &inBytes, &outputPtr, &outBytes);
    if (result == (size_t)-1) {
        iconv_close(cd);
        throw std::runtime_error("iconv conversion failed");
    }
    
    iconv_close(cd);
    
    return std::string(utf8Data.data(), utf8Data.size() - outBytes);
}
#else
[[nodiscard]] std::string convertBytesToString(const std::vector<std::byte>& byteData) {
    std::string str(byteData.size(), '\0');
    for (size_t i = 0; i < byteData.size(); ++i) {
        str[i] = static_cast<char>(byteData[i]);
    }

    return str;
}
#endif

void decompressDeflate(const std::vector<std::byte>& compressedData, std::vector<std::byte>& decompressedData, const uLongf decompressedSize) {
    z_stream strm{};
    strm.next_in = (Bytef*)compressedData.data();
    strm.avail_in = compressedData.size();
    strm.next_out = (Bytef*)decompressedData.data();
    strm.avail_out = decompressedSize;

    if (int ret = inflateInit2(&strm, -MAX_WBITS); ret != Z_OK) {
        throw std::runtime_error("inflateInit2 failed");
    }

    auto inflater = std::unique_ptr<z_stream, decltype(&inflateEnd)>(&strm, &inflateEnd);

    if (int ret = inflate(inflater.get(), Z_FINISH); ret != Z_STREAM_END) {
        throw std::runtime_error("inflate failed: " + std::to_string(ret));
    }

    decompressedData.resize(inflater->total_out);
}

void decompressData(const ZipHeader& header, const std::vector<std::byte>& compressedData, const std::string& outputFilePath) {
    std::vector<std::byte> decompressedData(header.uncompressed_size);
    
    switch(header.compression) {
        case 0: { // No Compression (Stored)
            spdlog::debug("Decompressing: {}. Compression Type: {}", outputFilePath, "No Compression (Stored)");
            decompressedData = compressedData;
            break;
        }
        case 8: { // Deflate
            spdlog::debug("Decompressing: {}. Compression type: {}", outputFilePath, "Deflate");
            decompressDeflate(compressedData, decompressedData, header.uncompressed_size);
            break;
        }
        default:
            throw std::runtime_error("Unsupported compression format.");
    }

    if (header.crc32_expected != 0) {
        const uLong calculatedCrc = crc32(0L, reinterpret_cast<const Bytef*>(decompressedData.data()), decompressedData.size());
        if (calculatedCrc != header.crc32_expected) {
            throw std::runtime_error("CRC32 check failed");
        }
    }

    std::ofstream outFile(outputFilePath, std::ios::binary);
    if (!outFile.write(reinterpret_cast<const char*>(decompressedData.data()), decompressedData.size())) {
        throw std::runtime_error("Failed to write decompressed data to file.");
    }
}

void decompressTask(const ZipHeader& header, const std::vector<std::byte>& compressedData, const std::string& outputFilePath) noexcept {
    try {
        decompressData(header, compressedData, outputFilePath);
    } catch (const std::exception& e) {
        spdlog::error("Error occurred: {}", e.what());
    }
}

void downloadAndDecompress(const std::vector<std::string>& zipList, const std::string& outpputDir) {
    ZipContentFetcher fetcher(zipList);
    std::vector<std::jthread> threads;
    
    while (true) {
        const auto signature = fetcher.read(4);
        if (signature->empty() || signature->size() != 4|| memcmp(signature->data(), "PK\x01\x02", 4) == 0) {
            spdlog::info("Finished reading file");
            break;
        }
        
        const auto header_data = fetcher.read(26);
        if (header_data->empty() || header_data->size() != 26) {
            spdlog::error("Error: Failed to read header data");
            break;
        }

        const ZipHeader header     = parseHeader(header_data->data());
        const auto fileNameData    = fetcher.read(header.file_name_length);
        const auto extra_field     = fetcher.read(header.extra_field_length);
        const auto compressData    = fetcher.read(header.compressed_size);
        const std::string fileName = convertBytesToString(*fileNameData);

        if (!compressData || compressData->size() != header.compressed_size) {
            spdlog::error("Error: Failed to read compressed data for file: {}", fileName);
            break;
        }

        const auto outputPath = std::filesystem::path(outpputDir) / fileName;
        if (fileName.back() == '/') {
            std::filesystem::create_directories(outputPath);
            continue;
        }
        
        threads.emplace_back([=, &fetcher]() {
            decompressTask(header, *compressData, outputPath.string());
        });
    }
}

[[nodiscard]] const std::string getCmdOption(const char** begin, const char** end, const std::string& longOption, const std::string& shortOption = "", const std::string& defaultValue = "") {
    auto itr = std::find_if(begin, end, [&](const char* arg) {
        return longOption == arg || (!shortOption.empty() && shortOption == arg);
    });

    if (itr != end && ++itr != end) {
        return *itr;
    }

    return defaultValue;
}

[[nodiscard]] bool cmdOptionExists(const char** begin, const char** end, const std::string& longOption, const std::string& shortOption = "") {
    return std::find_if(begin, end, [&](const char* arg) {
        return longOption == arg || (!shortOption.empty() && shortOption == arg);
    }) != end;
}


int main(int argc, const char* argv[]) {
    bool showHelp = cmdOptionExists(argv, argv+argc, "--help", "-h");
    if(showHelp){
        std::cout << 
R"(Zip Stream Downloader
usage: stream [-h] [--log-level {INFO,WARNING,ERROR,CRITICAL}] [--output-dir OUTPUT_DIR] [--urls URLS [URLS ...]]
        [--disable-crc32]

options:
  -h, --help            show this help message and exit
  --log-level {INFO,WARNING,ERROR,CRITICAL}
                        Sets the log level, default is `INFO`.
  --output-dir OUTPUT_DIR
                        Sets the download directory, default is `./output`.
  --urls URLS [URLS ...]
                        The list of URLs to download.
)" << std::endl;
        return 0;
    }

    const auto   logLevel = getCmdOption(argv, argv + argc, "--log-level", "-l", "INFO");
    if        (  logLevel == "DEBUG"    )   { spdlog::set_level(spdlog::level::debug)   ;}
    else if   (  logLevel == "INFO"     )   { spdlog::set_level(spdlog::level::info)    ;} 
    else if   (  logLevel == "WARNING"  )   { spdlog::set_level(spdlog::level::warn)    ;} 
    else if   (  logLevel == "ERROR"    )   { spdlog::set_level(spdlog::level::err)     ;} 
    else if   (  logLevel == "CRITICAL" )   { spdlog::set_level(spdlog::level::critical);} 
    else                                    { spdlog::warn("Unknown log level: {}", logLevel) ;}
    
    const std::string outputDir   = getCmdOption(argv, argv + argc, "--output-dir", "-o", "./output/");
    
    std::vector<std::string> urls;
    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--urls" || arg == "-u") {
            while (i + 1 < argc && argv[i + 1][0] != '-') {
                urls.push_back(argv[++i]);
            }
        }
    }

    if (urls.empty()) {
        spdlog::error("No URLs provided");
        return 1;
    }

    if(!std::filesystem::exists(outputDir)) {
        std::filesystem::create_directories(outputDir);
    }

    try {
        downloadAndDecompress(urls, outputDir);
    } catch (const std::exception& e) {
        std::cerr << "error: " << e.what() << std::endl;
    }
    
    return 0;
}

File: python/stream.py
# -*- coding: utf-8 -*-
import os
import re
import bz2
import sys
import lzma
import zlib
import struct
import argparse
from typing import List


import asyncio
import aiofiles
import requests
from loguru import logger

class ZipContentFetcher:
    def __init__(self, urls):
        self.urls = urls
        self.url_generator = self.url_generator_func()
        self.current_stream = None
        self.switch_stream()

    def url_generator_func(self):
        for url in self.urls:
            yield url

    def switch_stream(self):
        url = next(self.url_generator, None)
        if url is None:
            self.current_stream = None  # 所有URL都已处理完毕
        else:
            r = requests.get(url, stream=True)
            r.raise_for_status()
            self.current_stream = r.raw
            logger.info(f"开始处理分卷: {url}")

    def read(self, length):
        if self.current_stream is None:
            return b''  # 所有数据都已读取完毕

        data = self.current_stream.read(length)
        if len(data) < length:
            self.switch_stream()  # 当前文件已读完，切换到下一个
            logger.info(f"切换分卷, 偏移: {hex(length - len(data))}")
            data += self.read(length - len(data))
            
        return data


async def unzip(compression: bytes, compressData: bytes, file_name: str, output_file_path: str, crc32_expected: int):
    decompressed_data = b''
    if compression   == b'\x00\x00':  # No Compression (Stored)
        logger.info(f"正在解压: {file_name}, 格式: No Compression (Stored)")
        decompressed_data = compressData
    elif compression == b'\x08\x00':  # Deflate
        logger.info(f"正在解压: {file_name}, 格式: Deflate")
        decompressed_data = zlib.decompress(compressData, -zlib.MAX_WBITS)
    elif compression == b'\x0c\x00':  # BZIP2
        logger.info(f"正在解压: {file_name}, 格式: BZIP2")
        decompressed_data = bz2.decompress(compressData)
    elif compression == b'\x0e\x00':  # LZMA
        logger.info(f"正在解压: {file_name}, 格式: LZMA")
        decompressed_data = lzma.decompress(compressData)
    
    async with aiofiles.open(output_file_path, 'wb') as f:
        await f.write(decompressed_data)
        logger.info(f"{file_name} 解压完成")
        
    
    crc32_calculated = zlib.crc32(decompressed_data) & 0xffffffff
    if crc32_calculated == crc32_expected:
        logger.info(f"{file_name} CRC32校验成功.")
    else:
        logger.error(f"{file_name} CRC32校验失败. 计算的CRC32: {crc32_calculated}, 期望的CRC32: {crc32_expected}")


async def download(zip_list:List[str]):
    fetcher = ZipContentFetcher(zip_list)
    while True:
        signature = fetcher.read(4)
        if not signature or signature == b'PK\x01\x02':
            logger.info("文件读取结束")
            break
        
        header_data = fetcher.read(26)
        version, flags, compression, mod_time, mod_date, crc32_expected, compressed_size, uncompressed_size, file_name_length, extra_field_length = struct.unpack('<2s2s2s2s2sIIIHH', header_data)

        file_name = fetcher.read(file_name_length).decode('gbk', 'ignore')
        extra_field = fetcher.read(extra_field_length)
        
        logger.info(f"正在处理: {file_name}, 预期文件大小: {compressed_size}")
        compressData = fetcher.read(compressed_size)
        
        
        output_file_path = os.path.join(args.output_dir, file_name)
        if not os.path.exists(os.path.dirname(output_file_path)):
            os.makedirs(os.path.dirname(output_file_path))
            continue
        
        await unzip(compression, compressData, file_name, output_file_path, crc32_expected)


def parse_presets(game, download_type):
    game_id = {
        'bh3':'osvnlOc0S8',
        'genshin':'1Z8W5NHUQb',
        'sr':'64kMb5iAWu',
        'zzz':'x6znKlJ0xK'
    }
    data = requests.get(f"https://hyp-api.mihoyo.com/hyp/hyp-connect/api/getGamePackages?game_ids[]={game_id[game]}&launcher_id=jGHBHlcOq1").json()

    main = data["data"]["game_packages"][0]["main"]

    try:
        zip_list = main["major"]["game_pkgs"] if download_type == "major" else main["patches"][0]["game_pkgs"] # TODO: patch version
    except IndexError:
        logger.error("未找到指定类型的下载链接")
        sys.exit(1)

    return [pkg["url"] for pkg in zip_list]
      

def parse_args():
    parser = argparse.ArgumentParser(description="Zip Stream Downloader")
    parser.add_argument("--log-level", default="INFO", choices=["INFO", "WARNING", "ERROR", "CRITICAL"], help="设置日志级别，默认为 INFO ")
    parser.add_argument("--output-dir", default="./output", help="设置下载目录，默认为./output")
    parser.add_argument("--urls", nargs='+', help="需要下载的链接列表。")
    
    presets = ['genshin', 'zzz', 'bh3'] # , 'sr' 7z 暂不支持
    download_type = ['major', 'patch']
    parser.add_argument("--preset", default=None, choices=presets, help="使用预定义的下载地址。留空则需手动指定下载地址。")
    parser.add_argument("--type", default="major", choices=download_type, help="下载类型，默认为major.")

    return parser.parse_args()


logger.add("info.log", level='INFO', encoding='utf8')
logger.add("error.log", level='WARNING', encoding='utf8')

if __name__ == "__main__":
    args = parse_args()
    
    logger.remove()
    logger.add(sys.stdout, level=args.log_level.upper())
    logger.add("info.log", level=args.log_level.upper())
    
    if args.preset is not None:
        urls:List[str] = parse_presets(args.preset, args.type)
    else:
        if args.urls is None:
            logger.error("需要指定下载地址")
            sys.exit(1)
        urls = args.urls

    for url in urls:
        if url.endswith('.7z') or re.match(r'.+\.7z\.\d{3}',url):
            logger.error("暂不支持7z格式")
            sys.exit(1)
    
    url_str = "\r\n".join(urls)
    if input(f"{url_str} \r\n 以上为待下载列表，是否继续下载？(y/n): ") == 'n':
        sys.exit(0)
        
    if not os.path.exists(args.output_dir):
        os.makedirs(args.output_dir)

    asyncio.run(download(urls))
    
